### PR TITLE
🤖 feat: add deterministic stream guardrails for verification and doom loops

### PR DESCRIPTION
**Pull request by ibetitsmike**
## Summary

Adds two deterministic harness guardrails to the agent loop that enforce better agent behavior via the tool pipeline (not just prompting):

1. **Pre-completion verification guard** — gates `agent_report` to reject completion when the agent edited files but never ran any validation commands (tests, typecheck, lint). Allows through on a second attempt as an escape hatch.
2. **Doom-loop detection** — tracks per-file edit counts during a stream and injects a model-only nudge when the same file is edited 7+ times, telling the agent to step back and reconsider its approach.

### New per-stream tracker classes

- `StreamEditTracker` — counts edits per file path, supports one-time nudge per file per stream
- `StreamVerificationTracker` — tracks whether any validation-like bash commands were run, with a one-time nudge-then-allow-through lifecycle

Both are instantiated per-stream in `aiService.ts` and threaded through `ToolConfiguration` to tool factories.

### Verification guard (`agent_report`)

- Before returning `{ success: true }`, checks if edits occurred and no validation was attempted
- First attempt: throws with a clear error instructing the agent to run validation
- Second attempt: allows through (escape hatch for tasks where validation isn't applicable)
- Detection uses regex patterns matching common validation commands (`make test`, `bun test`, `vitest`, `tsc`, `run_and_report`, etc.)

### Doom-loop nudge (`file_edit_operation`)

- After each successful file write, records the edit in the tracker
- At threshold (7 edits to same file), attaches a `<notification>` via `__mux_notifications` (model-only, stripped before UI/persistence)
- Nudges once per file per stream (no spam)
- Skipped in plan-only mode

### Safety

- All behavior opt-in by tracker presence in `ToolConfiguration` — IPC tool calls without trackers see zero change
- Uses existing `__mux_notifications` infrastructure (already tested for stripping before persistence/UI)
- Conservative defaults: threshold 7, one nudge, one verification block then escape hatch

## Validation

- All new unit tests pass (`StreamEditTracker`, `StreamVerificationTracker`, `agent_report`, `bash`, `file_edit_operation`)
- `make typecheck` ✅
- `make lint` ✅
- `make fmt-check` ✅

---

_Generated with `mux` • Model: `anthropic:claude-opus-4-6` • Thinking: `xhigh` • Cost: `$1.44`_

### CODE REVIEW

**Comment by ibetitsmike** [comment]
@codex review

**Comment by ibetitsmike** [comment]
@codex review

Addressed feedback: tightened `run_and_report` detection to only match when the wrapped command itself is a validation command. Added negative test case for `run_and_report install bun install`.

**Comment by ibetitsmike** [comment]
@codex review

Addressed: new-file creation in `file_edit_insert.ts` now records edits via `editTracker.recordEdit()` in the create-file branch, ensuring both guardrails cover file creation workflows.

**Comment by ibetitsmike** [comment]
@codex review

Addressed: validation commands are now recognized after shell operators (`&&`, `||`, `;`, `|`), covering monorepo workflows like `cd packages/app && make test` and `run_and_report unit cd packages/app && bun test`.

**Comment by ibetitsmike** [comment]
@codex review

Addressed: removed `.*?` from `run_and_report` regex so only the actual command (third word) is checked. Added negative test for `run_and_report note echo make test`. Chained commands like `run_and_report unit cd app && bun test` still work because `&& bun test` is caught by the standalone pattern.

**Comment by ibetitsmike** [comment]
@codex review

Addressed with a clarifying code comment. Shell command parsing with regex is inherently imprecise — environment prefixes (`env CI=1`), shell wrappers (`bash -c`), etc., create an infinite regression of edge cases that regex alone can't solve. The `agent_report` escape hatch (second attempt always passes) is the designed safety net for any false negatives, adding at most one harmless retry. The current heuristic correctly handles the common cases (direct commands, `&&`/`;` chains, and `run_and_report` with direct validation commands).

**Comment by ibetitsmike** [comment]
Pausing the review loop here and leaving this for human direction.

**Re: bash-based edit tracking** — This was deliberately scoped out in the original plan. Tracking file edits via bash would require parsing arbitrary shell commands to figure out which files they write to (redirections, `sed`, `git`, formatters, etc.), which is fundamentally unreliable and a much larger, separate problem.

The `file_edit_*` tools are the **primary editing mechanism** for agents — the vast majority of code changes go through them. Covering those handles the common case. Bash-based edits are an inherent limitation that's acknowledged and can be addressed as a follow-up if data shows it's needed.

The alternative (heuristically guessing which files bash commands modify) would introduce a high false-positive rate and make the guardrail unreliable in the other direction.

**Comment by ibetitsmike** [comment]
@codex review

File: src/node/services/tools/agent_report.ts [hunk 1]
```diff
@@ -20,6 +20,20 @@ export const createAgentReportTool: ToolFactory = (config: ToolConfiguration) =>
         );
       }
 
+      // Guard: if edits were made but no validation was attempted, nudge the agent to verify.
+      const hasEdits = config.editTracker?.hasAnyEdits() ?? false;
+      const hasValidated = config.verificationTracker?.hasValidationAttempt() ?? false;
+      if (hasEdits && !hasValidated) {
```
[the remark this answers is not in the document]
Comment (ibetitsmike) [reply src/node/services/tools/agent_report.ts:26]: Intentionally out of scope — reliably detecting file modifications from arbitrary bash scripts (redirections, `sed -i`, `git checkout`, subshells, etc.) is not feasible with static regex. The escape hatch (second `agent_report` always passes) already covers this case: if an agent only edits files via bash and skips verification, the first rejection nudges them; the second attempt allows through. This keeps the guardrail simple and avoidable for legitimate use cases.

**Comment by ibetitsmike** [comment]
@codex review

File: src/node/services/tools/agent_report.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ibetitsmike) [reply src/node/services/tools/agent_report.ts:26]: Same design decision as the previous thread: reliably detecting file mutations from arbitrary bash scripts and git patches is infeasible with static analysis. The escape hatch (second `agent_report` always passes) handles this — if an agent only modifies files via bash/patches and doesn't run validation, the first rejection nudges it, the second allows through. This is the intended design tradeoff: catch the common case (tool-based edits) deterministically, handle edge cases (bash/patch edits) via the escape hatch.

File: src/node/services/tools/bash.ts [hunk 2]
```diff
@@ -522,6 +522,33 @@ function validateScript(script: string, config: ToolConfiguration): BashToolResu
...
+// Patterns for commands that are inherently validation (tests, typecheck, lint).
+// `run_and_report` is a generic wrapper (`run_and_report <step> <cmd...>`), so we
+// match it only when the wrapped command itself is a validation command.
+const VALIDATION_COMMAND_RE =
+  /(?:make\s+(?:test|typecheck|lint|static-check|fmt-check)|bun\s+(?:test|run\s+(?:test|typecheck|lint))|npm\s+(?:test|run\s+(?:test|typecheck|lint))|pnpm\s+(?:test|run\s+(?:test|typecheck|lint))|yarn\s+(?:test|run\s+(?:test|typecheck|lint))|tsc|eslint|vitest|pytest|cargo\s+(?:test|check|clippy))\b/;
+
+// Match validation commands at line start, after shell operators (&&, ||, ;, |),
+// or after run_and_report wrappers, to handle monorepo/subdirectory workflows
+// like `cd packages/app && make test`.
+const CMD_PREFIX = String.raw`(?:^|\n|&&|\|\||[;|])\s*`;
```
[the remark this answers is not in the document]
Comment (ibetitsmike) [reply src/node/services/tools/bash.ts:534]: Valid edge case, but the impact is a false *positive* (guard passes when it shouldn't), not a false *negative* (guard blocks when it shouldn't). In practice, agents rarely write heredocs containing validation-command text as their *only* interaction — real validation attempts produce actual command invocations. The heuristic is deliberately broad (false positives are acceptable, per the plan) since we only use it to avoid blocking `agent_report`, not to fail the run. Adding heredoc detection would require shell parsing complexity that isn't warranted for this edge case.

**Comment by ibetitsmike** [comment]
@codex review

File: src/node/services/tools/agent_report.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ibetitsmike) [reply src/node/services/tools/agent_report.ts:26]: Addressed in the same design as previous rounds: the escape hatch (second `agent_report` call always passes) handles cases where edits happen outside the tracked tool pipeline. Reliably detecting file mutations from arbitrary bash is infeasible. See previous thread discussions for full rationale.

File: src/node/services/tools/bash.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ibetitsmike) [reply src/node/services/tools/bash.ts:534]: Good catch — fixed in 95d45718c. Background commands (`run_in_background: true`) no longer count as validation attempts. Added tests verifying both foreground (marks) and background (doesn't mark) behavior.

**Comment by ibetitsmike** [comment]
@codex review

**Comment by ibetitsmike** [comment]
@codex review

File: src/node/services/tools/agent_report.ts [hunk 3]
```diff
@@ -20,6 +20,20 @@ export const createAgentReportTool: ToolFactory = (config: ToolConfiguration) =>
         );
       }
 
+      // Guard: if edits were made but no validation was attempted, nudge the agent to verify.
+      const hasEdits = config.editTracker?.hasAnyEdits() ?? false;
+      const hasValidated = config.verificationTracker?.hasValidationAttempt() ?? false;
```
[the remark this answers is not in the document]
Comment (ibetitsmike) [reply src/node/services/tools/agent_report.ts:25]: Same design decision as all previous rounds — intentionally out of scope. Escape hatch handles it. See earlier threads.

File: src/node/services/tools/bash.ts [hunk 4]
```diff
@@ -861,6 +888,11 @@ export const createBashTool: ToolFactory = (config: ToolConfiguration) => {
       const validationError = validateScript(script, config);
       if (validationError) return validationError;
 
+      // Mark validation attempts for the pre-completion verification guard.
+      if (looksLikeValidationCommand(script)) {
+        config.verificationTracker?.markValidationAttempt();
+      }
```
[the remark this answers is not in the document]
Comment (ibetitsmike) [reply src/node/services/tools/bash.ts]: Fixed in 07e9e67fc — background commands no longer mark verification.

File: src/node/services/tools/bash.ts [hunk 5]
```diff
@@ -861,6 +888,13 @@ export const createBashTool: ToolFactory = (config: ToolConfiguration) => {
       const validationError = validateScript(script, config);
       if (validationError) return validationError;
 
+      // Mark validation attempts for the pre-completion verification guard.
+      // Only count foreground commands — background processes haven't produced
+      // results yet, so they don't count as "validation attempted".
+      if (!run_in_background && looksLikeValidationCommand(script)) {
+        config.verificationTracker?.markValidationAttempt();
```
[the remark this answers is not in the document]
Comment (ibetitsmike) [reply src/node/services/tools/bash.ts:895]: Fixed in 07e9e67fc — background commands no longer mark verification.

File: src/node/services/tools/agent_report.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ibetitsmike) [reply src/node/services/tools/agent_report.ts:25]: Same design decision — intentionally out of scope. The escape hatch covers bash-based edits. See earlier threads for full rationale.

File: src/node/services/tools/agent_report.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ibetitsmike) [reply src/node/services/tools/agent_report.ts:26]: Good catch — fixed in f8b64a38f. File edits now call `verificationTracker.resetValidation()` so pre-edit validation no longer satisfies the guard. Added tests for the reset behavior.

File: src/node/services/tools/bash.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ibetitsmike) [reply src/node/services/tools/bash.ts:895]: Same intentional design decision — escape hatch covers bash-based edits. See earlier threads.

**Comment by ibetitsmike** [comment]
@codex review

File: src/node/services/tools/bash.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ibetitsmike) [reply src/node/services/tools/bash.ts:895]: Intentional design: we track *attempted* validation, not *successful* validation. If `cd missing && make test` fails, the agent sees the failure output and can fix + retry — the guardrail has already served its purpose by ensuring the agent thought about validation. The examples (`cd missing && make test`, `true || make test`) are edge cases that don't occur in practice; the escape hatch covers any remaining gaps. Moving validation marking to post-execution would add complexity for negligible benefit.

**Comment by ibetitsmike** [comment]
@codex review

File: src/node/services/tools/agent_report.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ibetitsmike) [reply src/node/services/tools/agent_report.ts:26]: This is the same suggestion that has been raised and resolved in 5+ previous review rounds. The design intentionally tracks only tool-based file edits; the escape hatch (second agent_report always passes) handles bash-based edits. See all previous threads for detailed rationale.

**Comment by ibetitsmike** [comment]
## Note for reviewers

Codex keeps raising the same P1/P2 about tracking bash-based file edits (redirections, `sed -i`, git commands, etc.) on every review round — this has been raised and resolved **6+ times** across review rounds.

**This is an intentional design decision:**
- Reliably detecting file mutations from arbitrary bash scripts is infeasible with static regex
- The escape hatch (second `agent_report` call always passes) handles this case
- The guardrail catches the common case (tool-based `file_edit_*` edits) deterministically
- Conservative defaults: one nudge, one block, then allow through

All other Codex feedback was addressed with code changes:
- ✅ Narrowed `run_and_report` regex to check actual wrapped command
- ✅ Tracked new-file creation in `file_edit_insert`
- ✅ Recognized validation commands after shell operators (`&&`, `;`, `|`)
- ✅ Excluded background commands from verification credit
- ✅ Reset validation state on file edits (pre-edit validation no longer counts)

CI checks are all passing. Ready for human review.

**Comment by ibetitsmike** [comment]
@codex review

File: src/node/services/tools/bash.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ibetitsmike) [reply src/node/services/tools/bash.ts:895]: Same as previous 6 rounds: intentional design. Validation is tracked as 'attempted' on regex match, not 'succeeded'. The escape hatch handles edge cases. This has been explained in detail in multiple resolved threads.

[the remark this answers is not in the document]
Comment (ibetitsmike) [reply src/node/services/tools/bash.ts:895]: Recurring suggestion (raised 7+ times). Intentional design decision — see PR comment and previous threads.

File: src/node/services/tools/agent_report.ts [hunk 6]
```diff
@@ -20,6 +20,20 @@ export const createAgentReportTool: ToolFactory = (config: ToolConfiguration) =>
         );
       }
 
+      // Guard: if edits were made but no validation was attempted, nudge the agent to verify.
+      const hasEdits = config.editTracker?.hasAnyEdits() ?? false;
```
[the remark this answers is not in the document]
Comment (ibetitsmike) [reply src/node/services/tools/agent_report.ts:24]: Recurring suggestion (raised 7+ times). Intentional design decision — see PR comment and previous threads.
